src/dbnode/x/encoding/shallow_cloner.go

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) 2024 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package encoding
+
+import (
+	"time"
+
+	enc "github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/dbnode/ts"
+	"github.com/m3db/m3/src/dbnode/x/xio"
+	xtime "github.com/m3db/m3/src/x/time"
+)
+
+// Cloner makes a clone of the given encoding type.
+type Cloner interface {
+	CloneSeriesIterator(enc.SeriesIterator) (enc.SeriesIterator, error)
+}
+
+type shallowCloner struct {
+	pools enc.IteratorPools
+
+	cloneBlockReaderFn cloneBlockReaderFn
+}
+
+// NewShallowCloner returns a cloner which makes a shallow copy of the given type.
+func NewShallowCloner(pools enc.IteratorPools) Cloner {
+	cloner := shallowCloner{
+		pools: pools,
+	}
+	cloner.cloneBlockReaderFn = cloner.cloneBlockReader
+	return cloner
+}
+
+// CloneSeriesIterator makes a shallow copy of the given series iterator.
+// i.e. The returned iterator can be iterated independently of the original iterator.
+// It does NOT copy the underlying tsz data, only points to the original iterator's data.
+// nb:
+//   - The lifecycle of the returned iterator is only valid until the original iterator is valid.
+//   - Do NOT iterate the original iterator, as it can release resources held on to by the cloned
```

Comment:
that's not strictly true, even a single Next() call (i.e. iterating once, not to the end) can call Close() - depending on how the blocks/iterators are laid out under the hood. 

i like the version of comments i have because they don't focus on the details of what's happening under the hood, they focus on giving callers guidelines to follow. if callers want to dig into why - they can go through the iterator code